[crm]: Add utility for CRM configuration

volodymyrsamotiy opened the pull request:
Signed-off-by: Volodymyr Samotiy <volodymyrs@mellanox.com>

Initial CRM  UI utility implementation for review. Please do not merge yet.

prsunny commented on crm/main.py:
```diff
@@ -0,0 +1,479 @@
+            for res in ["ipv4_route", "ipv6_route", "ipv4_nexthop", "ipv6_nexthop", "ipv4_neighbor", "ipv6_neighbor",
+                        "nexthop_group_member", "nexthop_group_object", "acl_table", "acl_group", "acl_entry",
+                        "acl_counter", "fdb_entry"]:
+                data.append([res, crm_info[res + "_threshold_type"], crm_info[res + "_low_threshold"], crm_info[res + "_high_threshold"]])
```
is crm_info[] from config_db prefixed with "crm" as mentioned in the design spec (CRM_IPV4_ROUTE_THRESHOLD_TYPE)?. Can you place the config_db output for understanding?

volodymyrsamotiy replied:
Currently "CONFIG_DB" output looks as following:
```
127.0.0.1:6379[4]> HGETALL CRM|Config
 1) "ipv4_route_threshold_type"
 2) "percentage"
 3) "ipv4_route_low_threshold"
 4) "70"
 5) "ipv4_route_high_threshold"
 6) "85"
...
```

prsunny commented on crm/main.py:
```diff
@@ -0,0 +1,479 @@
+        print tabulate(data, headers=header, tablefmt="simple", missingval="")
+        print '\n'
+
+    def show_acl_table_resources(self, table_id):
```
Show table resource requires user to specify table_id? Just thinking if we can also display the complete table. What do you say?

Was this intended in design spec, `[resources|thresholds] acl group [entry|counter]`

volodymyrsamotiy replied:
Currently it requires table_id, I agree that it is better to display complete table. Will check what we can do.

prsunny commented on crm/main.py:
```diff
@@ -0,0 +1,479 @@
+@click.argument('value', type=click.Choice(['percentage', 'used', 'free']))
+@click.pass_context
+def type(ctx, value):
+    """CRM threshod type configuration"""
```
Typo - threshold.. Please correct the ones below as well

prsunny commented:
Can you show one sample output for any one of the threshold and resource?

volodymyrsamotiy commented:
Yes, I will incorporate this script to the "show" and "config" utilities.

volodymyrsamotiy commented:
Please find "threshold" sample output below.
```
# crm show thresholds all


Resource Name         Threshold Type      Low Threshold    High Threshold
--------------------  ----------------  ---------------  ----------------
ipv4_route            percentage                     70                85
ipv6_route            percentage                     70                85
ipv4_nexthop          percentage                     70                85
ipv6_nexthop          percentage                     70                85
ipv4_neighbor         percentage                     70                85
ipv6_neighbor         percentage                     70                85
nexthop_group_member  percentage                     70                85
nexthop_group_object  percentage                     70                85
acl_table             percentage                     70                85
acl_group             percentage                     70                85
acl_entry             percentage                     70                85
acl_counter           percentage                     70                85
fdb_entry             percentage                     70                85
```

volodymyrsamotiy commented:
Please find "resource" sample output below.
```
# crm show resources all


Resource Name           Used Count    Available Count
--------------------  ------------  -----------------
ipv4_route                    6417              96925
ipv6_route                    6414              96925
ipv4_nexthop                     5              57314
ipv6_nexthop                     4              57314
ipv4_neighbor                    5              96925
ipv6_neighbor                    4              22114
nexthop_group_member             8              57314
nexthop_group_object             2              57314
fdb_entry                        1              96925




Stage    Bind Point    Resource Name      Used Count    Available Count
-------  ------------  ---------------  ------------  -----------------
INGRESS  PORT          acl_group                   0                399
INGRESS  PORT          acl_table                   0                393
INGRESS  LAG           acl_group                   0                399
INGRESS  LAG           acl_table                   0                393
INGRESS  VLAN          acl_group                   0                399
INGRESS  VLAN          acl_table                   0                393
INGRESS  RIF           acl_group                   0                399
INGRESS  RIF           acl_table                   0                393
INGRESS  SWITCH        acl_group                   0                399
INGRESS  SWITCH        acl_table                   0                393
EGRESS   PORT          acl_group                   0                399
EGRESS   PORT          acl_table                   1                393
...

```

lguohan commented:
can you change the name to ```nexthop_group``` since other object do not have this suffix.

```
nexthop_group_object             2              57314
```

lguohan commented:
can give the output for acl_entry/acl_counter for a table?

volodymyrsamotiy commented:
* Changed "show acl table" command in order to show complete table without passing table ID
* Renamed "nexthop_group_object" resource to "nexthop_group"
* Changed CRM polling interval represantation from minutes to seconds

volodymyrsamotiy commented:
Below is the output for the ACL per table resources: entry and counter.
(now ACL tables are also shown in "crm show resources all" command)
```
# crm show resources acl table


Table ID         Resource Name      Used Count    Available Count
---------------  ---------------  ------------  -----------------
0x70000000002bd  acl_entry                   1              13813
0x70000000002bd  acl_counter                 1              13813


```

prsunny commented on crm/main.py:
```diff
@@ -103,25 +104,33 @@ def show_acl_resources(self):
 
-        crm_stats = countersdb.get_all(countersdb.COUNTERS_DB, 'CRM:ACL_TABLE_STATS:{0}'.format(table_id))
+        # Retrieve all ACL table keys from CRM:ACL_TABLE_STATS
+        output = os.popen("docker exec -i database redis-cli --raw -n 2 KEYS *CRM:ACL_TABLE_STATS*").readlines()
```
In general, subprocess is being used instead of os.popen

prsunny commented on crm/main.py:
```diff
@@ -103,25 +104,33 @@ def show_acl_resources(self):
         countersdb.connect(countersdb.COUNTERS_DB)
 
-        header = ("Resource Name", "Used Count", "Available Count")
+        header = ("Table ID", "Resource Name", "Used Count", "Available Count")
```
Just checking, is it possible to display Table name as well, if it's available in the DB?

volodymyrsamotiy replied:
Looks like no, currently we cannot display ACL table name here.

qiluo-msft commented on crm/main.py:
```diff
@@ -0,0 +1,489 @@
+        header = ("Table ID", "Resource Name", "Used Count", "Available Count")
+
+        # Retrieve all ACL table keys from CRM:ACL_TABLE_STATS
+        proc = Popen("docker exec -i database redis-cli --raw -n 2 KEYS *CRM:ACL_TABLE_STATS*", stdout=PIPE, stderr=PIPE, shell=True)
```
It exec another script to make DB operation. Could you please use redis-py or existing XxxxDBConnector/etc to do DB changes?
